### DISCUSSION TITLE
Crazy idea: bi-directional sync with Photos

### DISCUSSION THREAD

**Discussion by RhetTbull** [RhetTbull/osxphotos#1233] [General]
I'm making good progress with [photokit](https://github.com/RhetTbull/photokit) which allows me to use the full private API to make changes to the Photos library. I took a small break to add iPhoto support to osxphotos....because I need to access an old iPhoto library. But I had a crazy thought -- it would be possible with the new code to create a "bi-directional" Photos sync.  That is, you could create an export of your Photos library then use another DAM (or just the file system) to manage it and edit photos. These changes would be detected by osxphotos and synced back to Photos. For example, you edit an exported image in Photoshop. The change is detected by osxphotos as an edit and it creates a change request in Photos to add the edit back to Photos.  You change a keyword on a photo and osxphotos detects that and adds it back to the original in Photos. You edit a photo in Photos and this is detected and the exported copy updated. This would work best with a background process that monitors for changes.  There's a lot of ways this could go wrong but it might be a way to "have your cake and eat it too" -- use Photos for syncing to iPhone, etc. but also be able to manage your photos "manually" in the Finder or using something like Digicam. Does this sound useful?

**Comment by cfc62** [comment]
So let me chime in here with a bit of background first...

While I'm knee-deep in the Apple ecosystem, I'm not 100% immersed. I do take a lot of pictures with my iPhones, but I also have an array of digital cameras; I tend to shoot RAW+JPEG most of the time, allowing me to at least try to recover an image if something goes wrong with the JPEG. I also shoot some video; also time-lapses from apps (iPhone and cameras) as well as from scratch. 

These all wind up in Photos, as that's how I tend to edit and cull, but I recognize as a personal shooter (as opposed to a professional photographer) I really don't have a post-processing workflow, and of course Photos is terrible for that. 

But Photos does work well and is likely the only choice for "consuming" (viewing) my library, whether that be in Photos directly or through a random screensaver or sharing with ATV4k for showing on my big screen OLED (OMG, freakin' wonderful PQ, it makes so many pictures I thought were poor look great). 

So, anything that would let me work with my photos (and videos) without having to export/reimport them in Photos would be a major plus. 

Yes, I have Photomator and RAW Power but I worry about them not keeping up with Photos schema changes, and I worry most about db corruption. In the end, I'd like to at least be able to consider LR, perhaps as my front end ingestion process.

I used to be concerned about the high and still growing number of background processes in MacOS, but I've come to accept it. 

TL;DR - yes please!

**Comment by hydrrrrr** [comment]
I agree with the post above. There are big improvements in raw processing software these days, especially with regard to noise reduction and enlargements. Also, we seem to be on the cusp of having support for 12bit HDR photo formats made available. I would love the ability to "replace" an old .JPEG with a new and better version,  that is viewed in 12bit HDR on Macs, IOS devices, Apple TVs etc. Imagine being able to do this without deleting and re-importing, tagging faces, etc. Just "update" the file with a new version. That would be incredible. 

I have no idea how difficult or complicated it be to create something like this, or if its possible at all. But osxphotos have surpassed every possible expectation by a mile so far. It is the best open source software I have installed on my Mac.

**Reply by RhetTbull** [reply]
It's possible but non-trivial. I think it would open up some interesting use cases if it could be implemented. A couple of limitations exist though: 1) changes via edits would not be hard -- e.g. you changed the file and that's done as an "edit" in Photos so you would now see the "revert to original" option when right clicking on Photo. That would preserve all metadata.  2) Replacing one file with another is also possible but this requires a behind the scenes export/re-import. In doing so, I cannot currently  preserve the tagged faces.  All other metadata can be preserved and I may be able to preserve the faces but currently there's no "official" way to do this and it would require some more reverse engineering. I'm working on a "replace in place" or "transform" option, see #909 to strip live videos, etc. and one option could be to replace the JPEG component of the RAW (I assume that's what you're referring to). That would be possible but again, currently faces would be lost.

**Reply by cfc62** [reply]
Even if you had to put some limitations* around such a replacement process, I think people would find value in it. 

* By limitations, I mean things like strict processes/steps to follow, losing certain metadata (or only saving a smaller set of metadata), only working via command line, etc. For most photos I think faces will be re-scanned, and even if some faces are lost, perhaps a tag could be added so that the user would know faces were lost and may need to be re-tagged.

**Reply by RhetTbull** [reply]
> perhaps a tag could be added so that the user would know faces were lost and may need to be re-tagged.

Yes, that would be easy. For example, osxphotos export has a `--person-keyword` option to write the person names to keywords. This would be an easy way to preserve this.
